graphql_compiler/schema_transformation/restrict_schema.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from copy import copy
+
+from graphql import build_ast_schema
+from graphql.language.visitor import REMOVE, Visitor, visit
+
+from ..ast_manipulation import get_ast_with_non_null_and_list_stripped
+from .utils import SchemaStructureError, get_query_type_name, get_scalar_names
+
+
+def restrict_schema(schema_ast, types_to_keep):
+    """Return new AST containing only a subset of types.
+
+    In addition to all types named in types_to_keep, the query type will also be kept. All user
+    defined scalars and enums will be kept.
+    All property fields will be kept, and only vertex fields that go to kept types will be
+    kept.
+
+    Args:
+        schema_ast: Document, representing the schema we're using to create a new schema with
+                    fewer types. It is not modified by this function
+        types_to_keep: Set[str], the set of names of types that we want to keep in the output
+                       schema
+
+    Returns:
+        Document, representing the schema that is derived from the input schema AST, but only
+        keeping a subset of types and a subset of vertex fields. The query type is additionally
+        also always kept
+
+    Raises:
+        SchemaStructureError if types_to_keep is inconsistent, in that some union type is included
+        but not all of its subtypes are included, or if some type has no remaining fields, or if
+        for some other reason the resulting schema AST cannot be made into a valid schema
+    """
+    schema = build_ast_schema(schema_ast)
+    query_type_name = get_query_type_name(schema)
+    scalar_names = get_scalar_names(schema)
+
+    visitor = RestrictSchemaVisitor(
+        types_to_keep, query_type_name, scalar_names
+    )
+    restricted_schema = visit(schema_ast, visitor)
+
+    try:
+        build_ast_schema(schema_ast)
+    except Exception as e:  # Can't be more specific, build_ast_schema throws Exceptions
+        raise SchemaStructureError(u'The resulting schema is invalid. Message: {}'.format(e))
+
+    # Note that it is possible for some types in the restricted schema to be unreachable
+    return restricted_schema
+
+
+class RestrictSchemaVisitor(Visitor):
+    """Remove types that are not explicitly kept, and fields go these types."""
+    def __init__(self, types_to_keep, query_type, scalars):
```

Comment:
I believe one of our linters is going to express a strong desire for a blank line between the class docstring and the init function.